24 - Rework/24/processed

antoniocaia opened the pull request:
Lo scopo di queste modifiche è quello di aumentare la modularità e seguire il principio di singola responsabilità , separando il parse del file da le operazione eseguite sui dati. 

Closes #24

antoniocaia committed:
Refactor (ProcessedFile): bow calcolate hours/day and hours/order

Processed file now calculate the total working hours for each working day and each order

antoniocaia committed:
Add (WrongFormatException): custom exception for wrong date/time format

antoniocaia committed:
Add (FormatChecker): collection of method to check different date/time format

Avoid duplicate code between different Parser

antoniocaia committed:
Update (Parser): method parse done

antoniocaia committed:
Fix (GestioneOrari ProcessedFile): minor problems

antoniocaia committed:
Refactor (ProcessedFile Parser): inverted dependency between ProcessedFile and Parser

Wabri commented on GestioneOrari/src/GestioneOrari.java:
```diff
@@ -1,25 +1,31 @@
 public class GestioneOrari {
 
-	public static void main(String[] args) {
+	public static void main(String[] args) throws ParseException, IOException {
```
Il main non deve lanciare eccezioni, dovresti usare il try catch per il metodo che lancia questi errori

Wabri commented on GestioneOrari/src/processed/ProcessedFile.java:
```diff
@@ -1,22 +1,69 @@
+
+	private @Getter HashMap<String, Float> orderHours;
+	private @Getter HashMap<Date, Float> dailyHours;
+	private @Getter float totalWorkOurs;
```
Typo

antoniocaia committed:
Fix (GestioneOrari): fixed main try-catch

antoniocaia committed:
Fix (ProcessedFile): fixed a typo and a bug when getting info

antoniocaia committed:
Add (Parser): added more control for time

Check that starting time is earlier then end time

antoniocaia committed:
Update: changed String type start and end hours into LocalTime

Using LocalTime allow to use default parser and help compairing and doing operation between instance.
Lost some flexibility with the input format!

antoniocaia committed:
Updare (src): moved src dir out of  GestioneOrari dir

Updated .gitignore

antoniocaia committed:
Add (FormatCheckerTest): add new test dir and tests

antoniocaia committed:
Fix (ProcessedFile): fixed bug with parser interface

antoniocaia committed:
Add (ProcessedFileTest): write test fro processedFile

Discovered and solved some bugs

antoniocaia committed:
Update (FormatChecker): add back the custom parser (currently not used)

MariusLovesPizza commented:
Domanda, i conflitti devono essere risolti ?

antoniocaia commented:
> Domanda, i conflitti devono essere risolti ?

Penso sia meglio aspettare le revisioni (e apportare le eventuali modifiche), e quando viene dato il via libera al merge allora si risolvono i conflitti.

MariusLovesPizza commented:
> 
> 
> > Domanda, i conflitti devono essere risolti ?
> 
> Penso sia meglio aspettare le revisioni (e apportare le eventuali modifiche), e quando viene dato il via libera al merge allora si risolvono i conflitti.

Quindi la sequenza è: revisionare, approvare, risoluzione conflitti, merge, delete branch

Wabri commented:
Non c'è una sequenza prestabilita. Di norma si fa la review quando lo sviluppatore toglie la pull request dal work in progress e di norma rimane in work in progress fintanto che le implementazioni non vengono completate e i conflitti risolti. Indipendentemente da questo la review la puoi fare sempre anche per aiutare l'altro sviluppatore. Usciti dalla wip (work in progress) si passa in code review, che può sia rimandare lo sviluppatore a cui è assegnata la pull request nuovamente in sviluppo perchè richieste delle modifiche oppure approvata e quindi dare la possibilità di effettuare la merge. Solo dopo che la merge viene completata si elimina il branch. ( @scrivereleonardo serve anche a te )

MariusLovesPizza commented:
> 
> 
> Non c'è una sequenza prestabilita. Di norma si fa la review quando lo sviluppatore toglie la pull request dal work in progress e di norma rimane in work in progress fintanto che le implementazioni non vengono completate e i conflitti risolti. Indipendentemente da questo la review la puoi fare sempre anche per aiutare l'altro sviluppatore. Usciti dalla wip (work in progress) si passa in code review, che può sia rimandare lo sviluppatore a cui è assegnata la pull request nuovamente in sviluppo perchè richieste delle modifiche oppure approvata e quindi dare la possibilità di effettuare la merge. Solo dopo che la merge viene completata si elimina il branch. ( @scrivereleonardo serve anche a te )

Ottima spiegazione. Bravo e grazie. 
Dove si vede che e' ancora in WiP oppure capire quando le implementazioni sono completate?

Wabri commented:
> Dove si vede che e' ancora in WiP oppure capire quando le implementazioni sono completate?

Ci sono diversi modi per vedere che è in wip:

- Se ci sono dei conflitti da risolvere
- Se è dichiaratamente indicato in wip, da poco github ha inserito una feature che permette di segnare come bozza la pr:
![2020-06-22-12-43-34_345x229](https://user-images.githubusercontent.com/12409541/85279222-5ac99880-b486-11ea-8177-f658d8aef082.png)
- Se non sono richiesti reviewers

MariusLovesPizza commented:
> 
> 
> > Dove si vede che e' ancora in WiP oppure capire quando le implementazioni sono completate?
> 
> Ci sono diversi modi per vedere che è in wip:
> 
>     * Se ci sono dei conflitti da risolvere
> 
>     * Se è dichiaratamente indicato in wip, da poco github ha inserito una feature che permette di segnare come bozza la pr:
>       ![2020-06-22-12-43-34_345x229](https://user-images.githubusercontent.com/12409541/85279222-5ac99880-b486-11ea-8177-f658d8aef082.png)
> 
>     * Se non sono richiesti reviewers

Good
Bene, questa linea l'avevo vista.

antoniocaia committed:
fix (format): fixed files format

antoniocaia committed:
Merge branch 'develop' into rework/24/Processed

antoniocaia commented:
> Non c'è una sequenza prestabilita. Di norma si fa la review quando lo sviluppatore toglie la pull request dal work in progress e di norma rimane in work in progress fintanto che le implementazioni non vengono completate e i conflitti risolti. Indipendentemente da questo la review la puoi fare sempre anche per aiutare l'altro sviluppatore. Usciti dalla wip (work in progress) si passa in code review, che può sia rimandare lo sviluppatore a cui è assegnata la pull request nuovamente in sviluppo perchè richieste delle modifiche oppure approvata e quindi dare la possibilità di effettuare la merge. Solo dopo che la merge viene completata si elimina il branch. ( @scrivereleonardo serve anche a te )

Grazie mille per la spiegazione, specialmente la parte dopo riguardo al wip.

antoniocaia commented:
Quando hai un po' di tempo @Wabri potresti dare uno sguardo al codice? Anche solo per sapere se hanno senso i cambiamenti che propongo o se conviene prendere un'altra strada.

Wabri commented on .gitignore:
```diff
@@ -2,4 +2,4 @@
 .settings/
 lib/
-bin/
+bin/
```
Alla fine di ogni file ci dovrebbe essere sempre EOF (end of file)  altrimenti non ci sono modi per capire se il file è concluso o no, nel 2020 non ci sono problemi però è teoricamente più corretto.

Wabri commented on src/GestioneOrari.java:
```diff
@@ -1,44 +1,38 @@
 	}
-
-}
+}
```
Stessa cosa del gitignore commentato sopra

Wabri commented on src/parser/FormatChecker.java:
```diff
@@ -0,0 +1,34 @@
+public class FormatChecker {
+
+	public static Date checkDate(String date) throws ParseException {
+		if (date.matches("(20[2-9][0-9])-([1-9]|0[1-9]|1[0-2])-([0-2][0-9]|3[0-1]|[1-9])"))
```
Cerca sempre di mettere le parentesi anche se c'è solo una sola istruzione. Serve a prevedere errori di scrittura.

Wabri commented on src/parser/FormatChecker.java:
```diff
@@ -0,0 +1,34 @@
+
+	public static LocalTime checkTimeCustomFormat(String time) throws WrongFormatException {
+		// TODO Bug: Edit String time in a format acceptable for LocalTime (HH:mm). Write tests!
+		if (time.matches("(0[0-9]|1[0-9]|2[0-3]|[0-9]):([0-5][0-9]|[0-9])"))
```
Stessa cosa dell'if del checkDate

Wabri commented on src/parser/FormatChecker.java:
```diff
@@ -0,0 +1,34 @@
+	}
+
+	public static void checkOrderTime(LocalTime start, LocalTime end) throws WrongFormatException {
+		if (!start.isBefore(end))
```
Stessa cosa dell'if di checkTimeCustomFormat

Wabri commented on src/parser/FormatChecker.java:
```diff
@@ -0,0 +1,34 @@
+		if (!start.isBefore(end))
+			throw new WrongFormatException("Start time must be earlier than end time. start: " + start + " end: " + end);
+	}
+}
```
Stessa cosa della classe GestioneOrari commentato sopra

Wabri commented on src/parser/Parser.java:
```diff
@@ -5,107 +5,48 @@
-		return oreLavorate / 60;
-	}
-}
+}
```
EOF

Wabri commented on src/parser/Parser.java:
```diff
@@ -5,107 +5,48 @@
-				Date date = checkDate(subStrings[0]);
-				String startHour = checkTime(subStrings[1]);
-				String endHour = checkTime(subStrings[2]);
+				Date date = FormatChecker.checkDate(subStrings[0]);
```
Questo ci voleva, non era compito di questa classe controllare il formato della data.

Wabri commented on src/parser/Parser.java:
```diff
@@ -5,107 +5,48 @@
 				String order = subStrings[3];
 				String note = (subStrings.length == 5) ? subStrings[4] : "";
-				processedFile.addRowToRows(new Row(date, startHour, endHour, order, note));
+				Row tmp = new Row(date, startHour, endHour, order, note);
```
Sarebbe meglio dargli un nome più significativo nonostante questo oggetto non abbia molti compiti

Wabri commented on src/parser/Parser.java:
```diff
@@ -5,107 +5,48 @@
 				rawRow = br.readLine();
 			}
-		} catch (WrongFormatException | ParseException e) {
+		} catch (ParseException | DateTimeParseException e) {
```
Attento a usare le condizioni sui tipi, rimane una good practice per evitare la duplicazione. @scrivereleonardo @MariusLovesPizza questa è una gestione del catch "avanzata" serve per evitare i try catch multipli rendendo il codice più leggibile e pulito.

MariusLovesPizza replied:
Si, anche a me piace di più usare un solo catch

> 
> 
> Attento a usare le condizioni sui tipi, rimane una good practice per evitare la duplicazione. @scrivereleonardo @MariusLovesPizza questa è una gestione del catch "avanzata" serve per evitare i try catch multipli rendendo il codice più leggibile e pulito.

Si, anche a me piace usare più questa versione di catch :D

antoniocaia replied:
Dovrei aver risolto (avevo dimenticato di catturare una eccezione), anche se non sono sicuro al 100% che fosse quello a cui ti riferivi

Wabri commented on src/processed/ProcessedFile.java:
```diff
@@ -1,22 +1,84 @@
+
+	public void updateData() throws IOException {
+		rows = parser.parse();
+		for (Row row : rows) {
```
@MariusLovesPizza @scrivereleonardo Questo costrutto si chiama **for each** anche questo permette di rendere il codice più pulito e leggero.

Wabri commented on src/processed/ProcessedFile.java:
```diff
@@ -1,22 +1,84 @@
+		return totalHours;
 	}
-}
+}
```
Ma non te li mette eclipse gli EOF?

antoniocaia replied:
E' più probabile li abbia tolti io inavvertitamente :)

Wabri commented on src/row/Row.java:
```diff
@@ -13,8 +14,8 @@
 	private String order;
 	private String note;
-}
+}
```
EOF

Wabri reviewed (changes requested):
Con questo refactoring il codice sembra più pulito e meno difficile da comprendere. 

Ottima gestione delle mappe e utilizzo dei costrutti.

PS: Avrei dovuto leggere prima i test, ma non li avevo visti quindi ho revisionato e commentato per ora solo le classi non test. Avevo pensato di provare a fare TDD una volta che si riesce a entrare in carreggiata con il progetto, ricordati di documentare le librerie che usi nel readme. Per ora dobbiamo fare così fintanto che non trasformiamo il progetto in maven.

antoniocaia committed:
Fix (FormatChecker): fixed brackets

antoniocaia committed:
Fix (Parser): fix try catch conditions

add WrongFormatException in try catch

antoniocaia committed:
Fix (Parser): fixed rrow name

name: tmp -> lineToInsert

antoniocaia committed:
Fix (format): fix EOF

antoniocaia commented:
> Con questo refactoring il codice sembra più pulito e meno difficile da comprendere.
> 
> Ottima gestione delle mappe e utilizzo dei costrutti.
> 
> PS: Avrei dovuto leggere prima i test, ma non li avevo visti quindi ho revisionato e commentato per ora solo le classi non test. Avevo pensato di provare a fare TDD una volta che si riesce a entrare in carreggiata con il progetto, ricordati di documentare le librerie che usi nel readme. Per ora dobbiamo fare così fintanto che non trasformiamo il progetto in maven.

Ammetto che i miei test rappresentano più un tentativo di test che test veri e propri... la teoria mi è abbastanza chiara ma metterla in pratica è un po' più dura. Quando riesco aggiorno il readme, ho usato mochito e assertj.

antoniocaia commented:
> > Con questo refactoring il codice sembra più pulito e meno difficile da comprendere.
> > Ottima gestione delle mappe e utilizzo dei costrutti.
> > PS: Avrei dovuto leggere prima i test, ma non li avevo visti quindi ho revisionato e commentato per ora solo le classi non test. Avevo pensato di provare a fare TDD una volta che si riesce a entrare in carreggiata con il progetto, ricordati di documentare le librerie che usi nel readme. Per ora dobbiamo fare così fintanto che non trasformiamo il progetto in maven.
> 
> Ammetto che i miei test rappresentano più un tentativo di test che test veri e propri... la teoria mi è abbastanza chiara ma metterla in pratica è un po' più dura. Quando riesco aggiorno il readme, ho usato mochito e assertj.

Una domanda: quanto è sbagliato definire getters da usare solo per i test?

antoniocaia committed:
Doc (README): instruction for mockito and assertj

Wabri commented:
> Una domanda: quanto è sbagliato definire getters da usare solo per i test?

Non è sbagliatissimo, i getters e setters li usi sempre. Diciamo che i test servono per coprire funzionalità non per crearne di nuove.

Wabri commented on README.md:
```diff
@@ -14,38 +15,79 @@ Considerare che in un file può esserci più di un giorno e che ogni giorno si p
-5. Creare una nuova cartella  `Rigth Click -> New -> Folder`
-6. Assegnare un nome alla cartella (es: *lib*) e cliccare su **Finish**.
-7. Aggiungere alla cartella *lib* appena creata il *lombok.jar* (Drag and Drop nell'IDEE, navigando nel filesystem, ... )
+7. Spostare  il *lombok.jar* nella cartella *lib* del progetto in cui vogliamo installare la dipendenza (Drag and Drop nell'IDEE, navigando nel filesystem, ... )
```
```suggestion
4. Spostare  il *lombok.jar* nella cartella *lib* del progetto in cui vogliamo installare la dipendenza (Drag and Drop nell'IDEE, navigando nel filesystem, ... )
```

Wabri commented on README.md:
```diff
@@ -14,38 +15,79 @@ Considerare che in un file può esserci più di un giorno e che ogni giorno si p
-6. Assegnare un nome alla cartella (es: *lib*) e cliccare su **Finish**.
-7. Aggiungere alla cartella *lib* appena creata il *lombok.jar* (Drag and Drop nell'IDEE, navigando nel filesystem, ... )
+7. Spostare  il *lombok.jar* nella cartella *lib* del progetto in cui vogliamo installare la dipendenza (Drag and Drop nell'IDEE, navigando nel filesystem, ... )
 8. Nell'IDEE selezionare nuovamente il progetto,  premere **F5** sulla propria tastiera oppure `Rigth Click -> Refresh`
```
```suggestion
5. Nell'IDEE selezionare nuovamente il progetto,  premere **F5** sulla propria tastiera oppure `Rigth Click -> Refresh`
```

Wabri commented on README.md:
```diff
@@ -14,38 +15,79 @@ Considerare che in un file può esserci più di un giorno e che ogni giorno si p
-7. Aggiungere alla cartella *lib* appena creata il *lombok.jar* (Drag and Drop nell'IDEE, navigando nel filesystem, ... )
+7. Spostare  il *lombok.jar* nella cartella *lib* del progetto in cui vogliamo installare la dipendenza (Drag and Drop nell'IDEE, navigando nel filesystem, ... )
 8. Nell'IDEE selezionare nuovamente il progetto,  premere **F5** sulla propria tastiera oppure `Rigth Click -> Refresh`
 9. Nella cartella *lib* dovrebbe essere comparso il *lombok.jar*. Selezionarlo e `Rigth Click -> Build Path -> Add to Build Path`
```
```suggestion
6. Nella cartella *lib* dovrebbe essere comparso il *lombok.jar*. Selezionarlo e `Rigth Click -> Build Path -> Add to Build Path`
```

Wabri commented on README.md:
```diff
@@ -14,38 +15,79 @@ Considerare che in un file può esserci più di un giorno e che ogni giorno si p
+7. Spostare  il *lombok.jar* nella cartella *lib* del progetto in cui vogliamo installare la dipendenza (Drag and Drop nell'IDEE, navigando nel filesystem, ... )
 8. Nell'IDEE selezionare nuovamente il progetto,  premere **F5** sulla propria tastiera oppure `Rigth Click -> Refresh`
 9. Nella cartella *lib* dovrebbe essere comparso il *lombok.jar*. Selezionarlo e `Rigth Click -> Build Path -> Add to Build Path`
 10. Chiudere e riaprire l'IDEE.
```
```suggestion
7. Chiudere e riaprire l'IDEE.
```

antoniocaia committed:
Update README.md

Co-authored-by: GabrielePuliti <gabriele.puliti@protonmail.com>

antoniocaia committed:
Update README.md

Co-authored-by: GabrielePuliti <gabriele.puliti@protonmail.com>

antoniocaia committed:
Update README.md

Co-authored-by: GabrielePuliti <gabriele.puliti@protonmail.com>

antoniocaia committed:
Update README.md

Co-authored-by: GabrielePuliti <gabriele.puliti@protonmail.com>
